### PR TITLE
GMP feature with lmdb

### PR DESCRIPTION
**Pull request by ray38**
some major updates that we have:

- GMP descriptors with primes
- single NN and corresponding scalers
- IO with parallel lmdbs (so they can be prepared in a parallel way)
- different caching strategy when training with lmdb

1. - no-caching (load directly from disk, very slow and inefficient, but support infinite amount of training data, not recommended)
2. - full-caching (load lmdbs to memory first, very fast IO and training, tho amount of training data limited by CPU RAM, recommended if possible, ~100x faster than the no-caching way)
3. - partial-caching (partially load lmdb into memory during training, close to full-caching speed but support infinite amount of training data. ~2-3x slower than full-caching, but ~30-50x faster than no-caching)

- relevant example files for preparing lmdb files, and training with them
- Note: I have commented out all the pca dim-reduction codes(do pca after fps are calculated, to reduce dimensionality) for now, but kept them there in case we want them in the future.

### CODE REVIEW

File: amptorch/descriptor/Gaussian/calculate_sf.cpp [hunk 1]
```diff
@@ -125,8 +127,10 @@ extern "C" int calculate_sf_cos(double** cell, double** cart, double** scale, in
     for (int i=0; i < 3; ++i) {
         // plane_d[i] / nbins[i] = height[i] / nbins[i] = size of bin in i dimension
         // cutoff * nbins[i] / plane_d[i] = cutoff / size of bin in i dimension = num bin needed in i dimension
+        // if cutoff < size: bin_range = 1, if cutoff > size: bin_range > 1
         bin_range[i] = ceil(cutoff * nbins[i] / plane_d[i]);
-        neigh_check_bins *= 2*bin_range[i];
+        // +1?
+        neigh_check_bins *= 2*bin_range[i]+1;
```
Comment (mshuaibii) [inline amptorch/descriptor/Gaussian/calculate_sf.cpp:133]: subtle - why did this change change?

File: amptorch/preprocessing/utils.py [hunk 2]
```diff
@@ -10,6 +13,92 @@
...
+                    # else:
+                    #     _values *= scale
+                    # _indices = data.fprimes._indices()
+                    # _size = data.fprimes.size()
+                    # data.fprimes = torch.sparse.FloatTensor(_indices, _values, _size)
+
+        return data_list
+
+
 class FeatureScaler:
```
Comment (mshuaibii) [inline amptorch/preprocessing/utils.py:102]: This looks fine - we should probably think of a cleaner way to do this or organize it. Not important for this PR, but something to keep in mind.

File: amptorch/trainer.py [hunk 3]
```diff
@@ -82,59 +94,99 @@ def get_unique_elements(self, training_images):
...
+            )
+            self.train_dataset = AtomsDataset(
+                images=training_images,
+                descriptor_setup=descriptor_setup,
+                forcetraining=self.forcetraining,
+                # pca_reduce=self.pca_reduce,
+                # pca_setting=self.config["dataset"].get(
+                #     "pca_setting",
+                #     {"num_pc": 20, "elementwise": False, "normalize": False},
+                # ),
```
Comment (mshuaibii) [inline amptorch/trainer.py:148]: Can  we remove these comments please. This can be added back at a later time for PCA customizability.

File: amptorch/trainer.py [hunk 4]
```diff
@@ -82,59 +94,99 @@ def get_unique_elements(self, training_images):
...
+            elif self.cache == "partial":
+                self.train_dataset = AtomsLMDBDatasetPartialCache(
+                    self.config["dataset"]["lmdb_path"],
+                )
+            elif self.cache == "no":
+                self.train_dataset = AtomsLMDBDataset(
+                    self.config["dataset"]["lmdb_path"],
+                )
+            else:
+                raise NotImplementedError
```
Comment (mshuaibii) [inline amptorch/trainer.py:112]: Any chance we can create a parent class for LMDBs that is instantiated with this `if/else` logic instead of having it all here?

File: amptorch/trainer.py [hunk 5]
```diff
@@ -82,59 +94,99 @@ def get_unique_elements(self, training_images):
...
             normalizers = {
                 "target": self.target_scaler,
                 "feature": self.feature_scaler,
             }
+            # save_normalizers(normalizers, os.path.join(self.cp_dir, "normalizers.json"))
             torch.save(normalizers, os.path.join(self.cp_dir, "normalizers.pt"))
+            # if self.pca_reduce:
+            #     torch.save(self.pca_reducer, os.path.join(self.cp_dir, "pca_reducer.pt"))
             # clean/organize config
+            # del self.config["dataset"]["fp_params"]
```
Comment (mshuaibii) [inline amptorch/trainer.py:171]: Let's remove all the commented out lines here

File: amptorch/trainer.py [hunk 6]
```diff
@@ -148,9 +200,17 @@ def load_model(self):
 
     def load_extras(self):
         callbacks = []
+        # check_memory_callback = check_memory()
+        # callbacks.append(check_memory_callback)
```
Comment (mshuaibii) [inline amptorch/trainer.py:204]: remove

File: amptorch/trainer.py [hunk 7]
```diff
@@ -252,8 +312,11 @@ def predict(self, images, disable_tqdm=True):
             cores=1,
         )
 
-        data_list = a2d.convert_all(images, disable_tqdm=disable_tqdm)
-        self.feature_scaler.norm(data_list, disable_tqdm=disable_tqdm)
+        data_list = a2d.convert_all(images, disable_tqdm=True)
```
Comment (mshuaibii) [inline amptorch/trainer.py:315]: `disable_tqdm=disable_tqdm` instead

File: amptorch/trainer.py [hunk 8]
```diff
@@ -252,8 +312,11 @@ def predict(self, images, disable_tqdm=True):
...
+
+        # if self.pca_reduce:
+        #     self.pca_reducer.reduce(data_list)
```
Comment (mshuaibii) [inline amptorch/trainer.py:318]: remove

File: amptorch/trainer.py [hunk 9]
```diff
@@ -252,8 +312,11 @@ def predict(self, images, disable_tqdm=True):
...
+        self.feature_scaler.norm(data_list)
```
Comment (mshuaibii) [inline amptorch/trainer.py:319]: `self.feature_scaler.norm(data_list, disable_tqdm=disable_tqdm)`

File: amptorch/trainer.py [hunk 10]
```diff
@@ -327,6 +390,10 @@ def load_pretrained(self, checkpoint_path=None, gpu2cpu=False):
             normalizers = torch.load(os.path.join(checkpoint_path, "normalizers.pt"))
             self.feature_scaler = normalizers["feature"]
             self.target_scaler = normalizers["target"]
+            # if self.config["dataset"].get("pca_reduce", False):
+            #     self.pca_reducer = torch.load(
+            #         os.path.join(checkpoint_path, "pca_reducer.pt")
+            #     )
```
Comment (mshuaibii) [inline amptorch/trainer.py:396]: remove

File: examples/construct_lmdb_load_scaler.py [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+                (dist * np.sin(0.65), dist * np.cos(0.65), 0),
+            ],
+        )
+        image.set_cell([10, 10, 10])
+        image.wrap(pbc=True)
+        image.set_calculator(EMT())
+        image.get_potential_energy()
+        images.append(image)
+
+    construct_lmdb(images, scaler_path, lmdb_path="./data2.lmdb")
```
Comment (mshuaibii) [inline examples/construct_lmdb_load_scaler.py:128]: Is this only difference between this and the other `construct_lmdb.py` is that this one assumes a normalizer is saved to disk and just loads that instead? If so, can we consolidate the 2 codes and just add an argument instead to do normalization or not.

File: amptorch/dataset_lmdb.py [hunk 12]
```diff
@@ -0,0 +1,365 @@
...
+            readonly=True,
+            lock=False,
+            readahead=False,
+            meminit=False,
+            max_readers=1,
+        )
+        return env
+
+
+class AtomsLMDBDatasetPartialCache(Dataset):
```
Comment (mshuaibii) [inline amptorch/dataset_lmdb.py:126]: For all the LMDB variants can you add a brief description under each one to highlight their differences please?

**Review by mshuaibii — CHANGES_REQUESTED** [review]
Looks good! Few comments and minor changes that would be good to address before we merge. Mostly organization.

I'll try playing around with the different LMDBs more closely to see if any specific optimizations/feedback there.

File: amptorch/preprocessing/atoms_to_data.py [hunk 13]
```diff
@@ -38,8 +38,18 @@ def convert(
...
+        descriptor_calculator = DescriptorCalculator(
+            images=[atoms],
+            descriptor=self.descriptor,
+            calc_derivatives=self.fprimes,
+            save_fps=self.save_fps,
+            cores=self.cores,
+            verbose=False,
+        )
+        self.descriptor_data = descriptor_calculator.prepare_descriptors()
+
```
Comment (mshuaibii) [inline amptorch/preprocessing/atoms_to_data.py:50]: This instantiates the descriptor calculator each time an atoms object is provided, whereas we can build the class once for a list of atoms object instead. Did the LMDB datasets require this to be moved here, if so do you remember why?

**Review by mshuaibii — COMMENTED** [review]
Can you test the following scenarios. The old LMDB PR didn't have the logic for this, but should be an easy fix if you haven't already done so:

1. Train a model with an LMDB. Load the checkpoint, and try retraining. Provide a config to the trainer class:
```
trainer_1 = AtomsTrainer(config_1)
trainer_1.train()
trained_cpdir = trainer_1.cp_dir
trainer_2 = AtomsTrainer(config_1)
trainer_2 .load_pretrained(trained_cpdir)
trainer_2 .train()
```

2. Train a model with an LMDB. Load the checkpoint, make a predictions. Provide **no** config to the trainer class: 
```
trainer_1 = AtomsTrainer(config_1)
trainer_1.train()
trained_cpdir = trainer_1.cp_dir
trainer_2 = AtomsTrainer()
trainer_2 .load_pretrained(trained_cpdir)
trainer_2 .predict([atoms_object])
```

Essentially an LMDB test of the following https://github.com/ulissigroup/amptorch/blob/master/amptorch/tests/pretrained_test.py

File: amptorch/descriptor/Gaussian/calculate_sf.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (ray38) [reply amptorch/descriptor/Gaussian/calculate_sf.cpp:133]: This is just to make it bullet-proof. it's basically allocating space based on the radius. the original way should be fine, but I'm a little bit afraid about extreme cases, since it doesn't include local row/column. Thus +1 in each dimension to include local row/column should make it bullet-proof.

File: amptorch/metrics.py [hunk 14]
```diff
@@ -96,18 +96,23 @@ def evaluator(
...
                 target_extractor=target_extractor,
             )
         )
-
+    callbacks.append(
+        Checkpoint(
+            monitor="{}_energy_{}_best".format(cp_on, metric),
+            fn_prefix="checkpoints/{}/".format(identifier),
+        )
+    )
```
[the remark this answers is not in the document]
Comment (ray38) [reply amptorch/metrics.py:112]: good idea! I'll see what I can do.

File: examples/construct_lmdb_load_scaler.py (hunk 11 above)
Comment (ray38) [reply examples/construct_lmdb_load_scaler.py:128]: yes, indeed. 
I kept them separate to make it more clear, cuz presumably the users need to run them separately. 
we can consolidate them tho if that makes more sense

File: amptorch/trainer.py (hunk 7 above)
Comment (ray38) [reply amptorch/trainer.py:315]: done

File: examples/construct_lmdb_load_scaler.py (hunk 11 above)
Comment (ray38) [reply examples/construct_lmdb_load_scaler.py:128]: consolidated
